gapic-generator/lib/gapic/resource_lookup.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Gapic
+  module ResourceLookup
+    def self.for_service service
+      file = service.parent
+      api = file.parent
+
+      resource_types = []
+      service.methods.map do |method|
+        resource_types += get_message_resource_types method.input
+      end
+
+      get_resources_descriptors_for api, *resource_types.flatten.compact.uniq
+    end
+
+    # @private
+    def self.get_message_resource_types message
+      resource_types = []
+      resource_types << message.resource.type if message.resource
+      resource_types += message.nested_messages.map { |nested_message| get_message_resource_types nested_message }
+      message.fields.each do |field|
+        resource_types << field.resource_reference.type if field.resource_reference
+        resource_types += get_message_resource_types field.message if field.message?
+      end
+      resource_types
+    end
+
+    # @private
+    def self.get_resources_descriptors_for api, *types
+      resources = []
+      api.files.each do |file|
+        resources += resources_descriptors_from_file file, *types
+        resources += file.messages.map { |m| resources_descriptors_from_message m, *types }
+      end
+      resources.flatten.compact.uniq.sort_by(&:type)
+    end
+
+    # @private
+    def self.resources_descriptors_from_file file, *types
+      resources = []
+
+      return resources if file.resources.nil?
+
+      resources += file.resources.select do |resource|
+        types.include? resource.type
+      end
+
+      resources
```

Comment:
This seems like it can be cleaned up:
```ruby
def self.resources_descriptors_from_file file, *types
  return [] if file.resource.nil?
  file.resources.select do |resource|
    types.include? resource.type
  end
end
```

Reply:
Oh yeah, this implementation should be greatly refactored. (And tested! (And passing rubocop!)) I didn’t want to spend time on it unless the approach was approved. Is the output correct?

Reply:
Oh, yes I think the output looks fine (modulo the one additional comment I made below about the redundant ArgumentError check).